File: tools/replay.py
"""Replays the fork quotes off-chain, in plain Python, and checks them to the wei.

The contract walks in integer arithmetic and draws its noise from keccak, so a
quote is a pure function of the block, the seed and the arguments. This script
re-implements that arithmetic (sdiv truncates toward zero, uint division
floors) and compares its mean and standard error with web/fork.js, which
test/Live.fork.t.sol wrote from the chain at a fixed block.

    python3 tools/replay.py        # needs pysha3 or pycryptodome for keccak
"""
import json, os, sys

try:
    from Crypto.Hash import keccak
    def k256(b):
        h = keccak.new(digest_bits=256); h.update(b); return h.digest()
except ImportError:
    import sha3
    def k256(b):
        return sha3.keccak_256(b).digest()

ONE = 10 ** 18
SEED = 1  # test/Live.fork.t.sol quotes with seed 1
MASK = 0x1FFFFF  # twelve 21-bit slices of one hash, Irwin-Hall


def sdiv(a, b):
    q = abs(a) // abs(b)
    return q if (a >= 0) == (b > 0) else -q


def normal(seed, p, t):
    h = int.from_bytes(k256(seed.to_bytes(32, "big") + ((p << 32) | t).to_bytes(32, "big")), "big")
    s = sum((h >> (21 * i)) & MASK for i in range(12))
    return sdiv((s - 6 * MASK) * ONE, 1 << 21)


def isqrt(x):
    if x == 0:
        return 0
    y, k = x, (x >> 1) + 1
    while k < y:
        y, k = k, (x // k + k) >> 1
    return y


def root(x):            # sqrt of a 1e18 fixed-point number, exact Babylonian
    return isqrt(x * ONE) if x else 0


def warm(x, guess):     # original arithmetic for the archived web/fork.js fixture
    if x == 0:
        return 0
    n, y = x * ONE, (guess or x)
    for _ in range(5):
        y = (y + n // y) >> 1
    return y


def exp_(x):            # e^x for 1e18 fixed point, Taylor on the fraction
    if x < -40 * ONE:
        return 0
    ax = -x if x < 0 else x
    n, fr = ax // ONE, ax % ONE
    term = ef = ONE
    for i in range(1, 13):
        term = term * fr // (ONE * i); ef += term
        if term == 0:
            break
    en = ONE
    for _ in range(n):
        en = en * 2718281828459045235 // ONE
    r = en * ef // ONE
    return ONE * ONE // r if x < 0 else r


def quote(f, steps, every, paths, seed):
    spot, om, al, be, v0 = f["spot"], f["omega"], f["alpha"], f["beta"], f["varStep"]
    sd0, pays = root(v0), []
    for p in range(paths):
        v, sd, logS, acc = v0, sd0, 0, 0
        for t in range(steps):
            r = sdiv(sd * normal(seed, p, t), ONE)
            logS += r - v // 2
            v = om + al * sdiv(r * r, ONE) // ONE + be * v // ONE
            sd = warm(v, sd)
            if (t + 1) % every == 0:
                acc += exp_(logS)
        avg = spot * (acc // (steps // every)) // ONE
        pays.append(avg - spot if avg > spot else 0)
    n = paths
    mean, e2 = sum(pays) // n, sum(x * x for x in pays) // n
    var = (e2 - mean * mean) * n // (n - 1) if e2 > mean * mean else 0
    return mean, isqrt(var // n)


def main():
    here = os.path.dirname(os.path.abspath(__file__))
    s = open(os.path.join(here, "..", "web", "fork.js")).read()
    d = json.loads(s[s.index("{"):])
    print(f"block {d['block']}, seed {SEED}, {d['paths']} paths, {d['steps']} steps")
    ok = True
    for f in d["feeds"]:
        mean, se = quote(f, d["steps"], d["every"], d["paths"], SEED)
        hit = (mean, se) == (f["asianCall"], f["se"])
        ok &= hit
        print(f"{f['symbol']:>5}  python {mean} +- {se}\n       chain  {f['asianCall']} +- {f['se']}  {'MATCH' if hit else 'MISMATCH'}")
    print("all nine match to the wei" if ok else "MISMATCH")
    sys.exit(0 if ok else 1)


if __name__ == "__main__":
    main()

File: web/replay.js
// The contract's walk in JavaScript, so a browser can replay a quote to the wei
// from the spot, variance and seed the quote returned. tools/replay.py is the
// same arithmetic in Python; test/Live.fork.t.sol wrote the numbers both are
// checked against. Division of a negative number truncates toward zero, as in
// the EVM; every other division is of non-negative numbers and floors.
const ONE = 10n ** 18n, MASK = 0x1FFFFFn, E = 2718281828459045235n;

const sdiv = (a, b) => { const q = (a < 0n ? -a : a) / (b < 0n ? -b : b); return (a >= 0n) === (b > 0n) ? q : -q; };

function normal(k256, seed, p, t) {
  const buf = new Uint8Array(64);
  let s = seed; for (let i = 31; i >= 0; i--) { buf[i] = Number(s & 0xFFn); s >>= 8n; }
  let v = (BigInt(p) << 32n) | BigInt(t); for (let i = 63; i >= 32; i--) { buf[i] = Number(v & 0xFFn); v >>= 8n; }
  const h = BigInt("0x" + k256(buf));
  let sum = 0n; for (let i = 0; i < 12; i++) sum += (h >> BigInt(21 * i)) & MASK;
  return sdiv((sum - 6n * MASK) * ONE, 1n << 21n);
}

function isqrt(x) { if (x === 0n) return 0n; let y = x, k = (x >> 1n) + 1n; while (k < y) { y = k; k = (x / k + k) >> 1n; } return y; }
const root = (x) => (x ? isqrt(x * ONE) : 0n);
function warm(x, g) { if (x === 0n) return 0n; const n = x * ONE; let y = g || x; for (let i = 0; i < 5; i++) y = (y + n / y) >> 1n; return y; }
function exp_(x) {
  if (x < -40n * ONE) return 0n;
  const ax = x < 0n ? -x : x, n = ax / ONE, fr = ax % ONE;
  let term = ONE, ef = ONE;
  for (let i = 1n; i <= 12n; i++) { term = term * fr / (ONE * i); ef += term; if (term === 0n) break; }
  let en = ONE; for (let i = 0n; i < n; i++) en = en * E / ONE;
  const r = en * ef / ONE;
  return x < 0n ? ONE * ONE / r : r;
}

/** Replays the at-the-money average-price call. k256 maps bytes to a hex digest. */
export function replay(k256, { spot, varStep, omega, alpha, beta, seed, steps, every, paths }) {
  const sd0 = root(varStep), marks = BigInt(Math.floor(steps / every));
  const pays = [], series = [];
  for (let p = 0; p < paths; p++) {
    let v = varStep, sd = sd0, logS = 0n, acc = 0n; const path = [];
    for (let t = 0; t < steps; t++) {
      const r = sdiv(sd * normal(k256, seed, p, t), ONE);
      logS += r - v / 2n;
      v = omega + alpha * sdiv(r * r, ONE) / ONE + beta * v / ONE;
      sd = warm(v, sd);
      const e = exp_(logS);
      path.push(Number(spot * e / ONE) / 1e18);
      if ((t + 1) % every === 0) acc += e;
    }
    const avg = spot * (acc / marks) / ONE;
    pays.push(avg > spot ? avg - spot : 0n); series.push(path);
  }
  const n = BigInt(paths); let sum = 0n, e2 = 0n;
  for (const x of pays) { sum += x; e2 += x * x; }
  const mean = sum / n, m2 = mean * mean; e2 /= n;
  const vr = e2 > m2 ? (e2 - m2) * n / (n - 1n) : 0n;
  return { mean, se: isqrt(vr / n), series, pays };
}
